### ISSUE TITLE
manifestdir is deprecated in Puppet 3.6.0

### ISSUE THREAD

**Issue by ferventcoder** [mitchellh/vagrant#3740]
manifest now has the ability to take a directory or fully qualified path to a manifest file. If it is a directory it will parse all files in there and run the result.

Heads up what you are going to start seeing issues on:

==> default: Running provisioner: puppet...
Running Puppet with /tmp/vagrant-puppet-4/manifests/site.pp...
Warning: Setting manifestdir is deprecated. See http://links.puppetlabs.com/env-settings-deprecations
   (at c:/Program Files/Puppet Labs/Puppet/puppet/lib/puppet/settings.rb:260:in `block (2 levels) in parse_global_options'; c:/Program Files/Puppet Labs/Puppet/puppet/lib/puppet/util/command_line/puppet_option_parser.rb:81:in `call'; c:/Program Files/Puppet Labs/Puppet/puppet/lib/puppet/util/command_line/puppet_option_parser.rb:81:in `block in pass_only_last_value_on_to'; c:/Program Files/Puppet Labs/Puppet/puppet/lib/puppet/util/command_line/trollop.rb:432:in `call'; c:/Program Files/Puppet Labs/Puppet/puppet/lib/puppet/util/command_line/trollop.rb:432:in `block in parse'; c:/Program Files/Puppet Labs/Puppet/puppet/lib/puppet/util/command_line/trollop.rb:393:in `each'; c:/Program Files/Puppet Labs/Puppet/puppet/lib/puppet/util/command_line/trollop.rb:393:in `parse'; c:/Program Files/Puppet Labs/Puppet/puppet/lib/puppet/util/command_line/puppet_option_parser.rb:74:in `parse'; c:/Program Files/Puppet Labs/Puppet/puppet/lib/puppet/settings.rb:270:in `parse_global_options'; c:/Program Files/Puppet Labs/Puppet/puppet/lib/puppet/settings.rb:239:in `initialize_global_settings'; c:/Program Files/Puppet Labs/Puppet/puppet/lib/puppet.rb:150:in `do_initialize_settings_for_run_mode'; c:/Program Files/Puppet Labs/Puppet/puppet/lib/puppet.rb:136:in `initialize_settings'; c:/Program Files/Puppet Labs/Puppet/puppet/lib/puppet/util/command_line.rb:86:in `block in execute'; c:/Program Files/Puppet Labs/Puppet/puppet/lib/puppet/util.rb:479:in `exit_on_fail'; c:/Program Files/Puppet Labs/Puppet/puppet/lib/puppet/util/command_line.rb:85:in `execute'; c:/Program Files/Puppet Labs/Puppet/puppet/bin/puppet:4:in `<main>')

I'm thinking we can change the provisioner from this:
```ruby
 config.vm.provision :puppet, :options => ["--debug --trace --verbose"] do |puppet|
    puppet.manifests_path = "puppet/manifests"
    puppet.manifest_file  = "site.pp"
    puppet.module_path = "puppet/modules"
  end
```
to this:
```ruby
 config.vm.provision :puppet, :options => ["--debug --trace --verbose"] do |puppet|
    puppet.manifest_file  = "puppet/manifests/site.pp"
    puppet.module_path = "puppet/modules"
  end
```
or try something else? Thoughts?

From my comments below, leave the Ux the same and merge the options together and provide them to puppet as one setting.

**Comment by netmikey** [comment]
+1

**Comment by zipkid** [comment]
Hello.... 
Puppet 4 is about to be released quite soon.
I am in the process of testing this and because of this i can not use the puppet provisioner.

  ==> puppetserver: Running provisioner: puppet...
  ==> puppetserver: Running Puppet with default.pp...
  ==> puppetserver: Error: Could not parse application options: invalid option: --manifestdir

It seems like the discussion of how to fix/implement this is done, so it would be extremely useful to have this fixed real soon...

Stefan - Zipkid - Goethals.

**Comment by zipkid** [comment]
@benh57 , did you manage to get some code ready to submit for this?

**Comment by laapsaap** [comment]
+1

We cant test puppet 4 rc2 because of this with vagrant.

**Comment by benh57** [comment]
Will continue work on this this weekend.

**Comment by benh57** [comment]
If anyone would like to test it out, code currently is on my fork:

https://github.com/benh57/vagrant/tree/environments_wip
You can specify:
       puppet.environment_path = "../puppet/environments"
       puppet.environment = "testenv"

-- If only environment and environments_path are specified, it will parse and use the manifest specified in the environment.conf file.
-- If manifests_path and manifest_file is specified along with the environment options, the manifest from the environment will be overridden by the specified manifest_file.
-- If manifests_path and manifest_file are specified without environments, the old non-environment mode will be used (which will fail on Puppet 4+)

Will submit PR soon.

**Comment by laapsaap** [comment]
^^ Confirmed working https://github.com/benh57/vagrant/tree/environments_wip

Its working great. Thank you!

**Comment by roman-mueller** [comment]
Puppet 4 has been released and now Vagrant won't work anymore without shell provisioner workarounds. 
Would be great to get this solved soon.

**Comment by l0b0** [comment]
Is there a *configuration* workaround for [masterless Puppet](https://github.com/l0b0/root) (that is, without installing a fork or older version)?

    $ pacman --query --owns "$(which puppet)"
    /usr/bin/puppet is owned by puppet 4.1.0-1
    $ vagrant --version
    Vagrant 1.7.2

**Comment by ball-hayden** [comment]
I think the only workaround that wouldn't involve forking or using an older version of puppet would be to run a shell provisioner that invokes puppet apply with the correct flags...

**Comment by ball-hayden** [comment]
Although, @sethvargo has closed this issue - has this been fixed on master?

**Comment by sethvargo** [comment]
@ball-hayden Puppet 4 support has landed on master, but is not yet in a released version. Thanks!

**Comment by lmayorga1980** [comment]
@ball-hayden Fixed on master but you will get a PATH issue.

![screen shot 2015-06-25 at 11 18 46 am](https://cloud.githubusercontent.com/assets/3172378/8358052/21fc287c-1b2c-11e5-81a2-42ef89ef64eb.png)

**Comment by benh57** [comment]
We can fix the puppet agent path autodetection by adding more common paths to check.

But, to work around you can use the new config parameter 'puppet.binary_path' to pass the path to your puppet agent.

**Comment by lmayorga1980** [comment]
:+1: @benh57

**Comment by ferventcoder** [comment]
@lmayorga1980 what was that issue again? I provided what I think is the commit that causes the issue.

**Comment by ferventcoder** [comment]
Found it #5806

**Comment by lmayorga1980** [comment]
@ferventcoder seems like my comment was just noise. I think there is another issue for this.

**Comment by martijn** [comment]
So Vagrant 1.7.4 still calls puppet with --manifestdir, yet newer boxes like centos-7.0-64-puppet 1.0.2 include a version of Puppet that no longer supports this parameter. Seems like that needs fixing.

**Comment by thenexxuz** [comment]
This helped me at least get up and running without needing to do much work on updating a lot of things.

I added this to my Vagrantfile after the `config.vm.box` line:
`config.vm.box_version = "1.0.1"`

**Comment by codesplicer** [comment]
+1 
Still facing this issue with ```Error: Could not parse application options: invalid option: --manifestdir``` using the ```puppetlabs/ubuntu-12.04-64-puppet``` box, which ships with puppet v4.2.1.

**Comment by jbeard6** [comment]
For what it's worth, I have worked around this issue by using a script provisioner to invoke puppet apply directly.  This not only has the benefit of working, but also more closely replicates how puppet will actually be used in a production environment to apply my config.

Still, I'm looking forward to having this issue resolved.  This has been a long time coming.

**Comment by ball-hayden** [comment]
I have been working around this by placing the contents of https://github.com/benh57/vagrant/tree/environments_wip/plugins/provisioners/puppet into `lib/puppet` in my project directory, adjusting the plugin name to `puppet4`, then adding `require_relative 'lib/puppet/plugin.rb'` at the top of my `Vagrantfile`.

**Comment by benh57** [comment]
With puppet 4 you need to switch your Vagrantfile to use puppet environments, instead of the manifest file setting. This is documented in the vagrant docs.  I'm not sure how that premade box is set up, but i can try it out.

**Comment by benh57** [comment]
* One would assume if you pass a path to manifest file, that you want to use that manifest file (and the old option). 

I suppose i could, if an environment file is not specified, make a 'fake' environment file, pointing at the specified manifest file. However, that would only work going back a couple versions of puppet, i think it would work on 3 or later? We have no autodetection of puppet version, but that could potentially be added.

**Comment by benh57** [comment]
@ball-hayden I think all of that branch ('environments_wip' ) made it into Vagrant 1.7.4, didn't it? Maybe it was broken later by another change?

**Comment by benh57** [comment]
@jbeard6 It was fixed. If you're using puppet 4, pass your environment. This will put it into environments mode.

    puppet.environment = "production"
    puppet_environment_path = "path/to/environments/on/host"
(and optionally, the path to a manifest - which will override the environment's manifest if you set it!)
      puppet.manifests_path = "puppet/manifests"
      puppet.manifest_file = "site.pp"

If you JUST to the last 2 settings, we *assume* that you are on puppet <4, and pass --manifestdir.

In other words -- this bug is not a bug anymore.

**Comment by codesplicer** [comment]
@benh57 This solution assumes that you are in fact using puppet environments with a puppet master, which isn't the case for those of us operating puppet masterless.

**Comment by codesplicer** [comment]
For the record I managed to workaround this issue by using symlinks. Given the following folder tree:

```
project_foo
    \ puppet
        | puppet.conf
        \ modules
        \ manifests
```

I created a new folder ```environments```which contained a symlink to the puppet folder in the ```project_foo``` folder

```
project_foo
    \ puppet
    | puppet.conf
        \ modules
        \ manifests
        \ environments
            \ production    <-- Symlink to ../../
```

Then in my Vagrantfile, as suggested by @benh57 I added the following:
```
config.vm.provision "puppet" do |puppet|
    puppet.environment = "production"
    puppet.environment_path = "puppet/environments"
    ...
```

Its a somewhat dirty hack but it works for now.

**Comment by adamcstephens** [comment]
As can be seen by the removal of manifestdir (following the original deprecation notice), environments are intended for both puppet apply and agent.  vagrant has support for this now, so I agree with @benh57 that this is no longer a bug.

See _BREAK: Directory Environments Replace Config File Environments_ here: http://docs.puppetlabs.com/puppet/4.0/reference/release_notes.html

**Comment by l0b0** [comment]
@adamcstephens From reading [Directory Environments Replace Config File Environments](https://docs.puppetlabs.com/puppet/4.0/reference/release_notes.html#break-directory-environments-replace-config-file-environments) and the linked documentation I don't see how it's *possible* to use environments with masterless Puppet. Some relevant quotes:

[Creating Environments](https://docs.puppetlabs.com/puppet/latest/reference/environments_creating.html):
> [The environment] must be located on the Puppet master server(s)

What if I don't have a master at all, even less a separate server? There doesn't seem to be any mention of masterless Puppet in this article.

[About Environments](https://docs.puppetlabs.com/puppet/latest/reference/environments.html):
> A Puppet master server can serve each environment with completely different main manifests and modulepaths.
> [...]
> In Puppet manifests, you can get the name of the current environment by using the $environment variable, which is set by the Puppet master.

Again, what about masterless Puppet?

[`$environment` configuration reference](https://docs.puppetlabs.com/references/4.2.latest/configuration.html#environment):
>  For clients (e.g., `puppet agent`) this determines the environment itself, which is used to find modules and much more. For servers (i.e., `puppet master`) this provides the default environment for nodes we know nothing about.

No mention of `puppet apply`. In fact all the documentation around environments seems to assume a master/agent setup, and doesn't mention (in any way I have been able to discover) how it applies to masterless Puppet.

**Comment by benh57** [comment]
I use vagrant with my environments all day with no puppet master, to test
my configurations for (later) use with the puppet master server.

It's just a named folder, with an environment.conf file in it. I'm not sure
how that's a hack, since it's the intended puppet  (4) way.

Maybe it would be 'cleaner' if vagrant would detect your puppet version and
'just work' and help you out by 'faking' an environment ? That seems almost
more of a hack. Though it would have the benefit that existing Vagrantfiles
would continue to work. Perhaps vagrant could warn..

**Comment by sultano** [comment]
@adamcstephens @benh57 It would seem @l0b0 may have a valid point. I agree the intended Puppet 4 way would be to use directory environments instead of dynamic environments.

> Starting with Puppet 3.6, Directory Environments started taking over from Dynamic Environments as Puppet’s mechanism for serving different versions of modules and code. In Puppet 4, they’re the default and other environment support is gone.

Source: [Directory Environments Replace Config File Environments](https://docs.puppetlabs.com/puppet/4.0/reference/release_notes.html#break-directory-environments-replace-config-file-environments)

I just fail to see why the environment needs to be specified when using `puppet apply`

> Puppet apply does not use the manifest from an environment; it always uses the manifest given on the CLI.

Source: [Directories: The Main Manifest(s)](https://docs.puppetlabs.com/puppet/latest/reference/dirs_manifest.html#with-puppet-apply)

So Vagrantfile's without the environment specified in a masterless setup should still work as expected?

**Comment by sultano** [comment]
As a workaround, I've set the following in my Vagrantfile to force environments mode without having to change the folder structure.
```
puppet.environment = "production"
puppet.environment_path = "../../"
```

**Comment by benh57** [comment]
Suleman -- I chose to have it default to using the old system, so that pre
existing vagrantfiles + machines would continue to work.

Vagrant currently has no mechanism to detect your puppet version.

If you're on the older puppet versions, no changes need be made. Changes to
vagrantfiles are only required if you upgrade puppet. I don't consider
specifying the environment as you are doing a workaround, rather it's the
intended behavior.

Ideal would probably be if the mode flipped automatically depending on your
installed puppet version -- anyone is free to implement that..  I just
implemented this based on my needs, well after posting the proposal here.
I'm not a core vagrant developer.

-Ben

On Wed, Oct 14, 2015 at 2:19 AM, Suleman Chikhalia <notifications@github.com

**Comment by ferventcoder** [comment]
Vagrant prolly could query Puppet to see what version it is and adjust accordingly.

**Comment by martijn** [comment]
Ok I missed the whole environments / Puppet 4 thing (should've read the changelog). I'm going to try it out but I assume that fixes the issue. Thanks for clarifying @benh57

**Comment by iainhallam** [comment]
The documentation currently says (https://www.vagrantup.com/docs/provisioning/puppet_apply.html):

> If you are using Puppet 4 or higher, you can provision using Puppet Environments by specifying the name of the environment and the path on the local disk to the environment files

I read this and thought that since I was using only a single manifest, I wouldn't use the environments configuration as it's just extra things to do, when all I want is to apply a single file. If Vagrant is actually using environment configuration as a flag to decide whether to use Puppet 4-style calling, then this needs to be changed to "must".

An alternative would be to add a "manifest" configuration that also makes Vagrant use v4-style calls to Puppet but can take a path and filename together.
